src/Components/test/testassets/BasicTestApp/RouterTest/TestRouterWithDynamicAssembly.razor

```diff
@@ -0,0 +1,56 @@
+@using Microsoft.AspNetCore.Components.Routing
+@using Microsoft.AspNetCore.Components.WebAssembly
+@using System.Linq
+@using System.Reflection
+
+@inject WebAssemblyDynamicAssemblyLoader lazyLoader
+
+<Router AppAssembly="@typeof(BasicTestApp.Program).Assembly" AdditionalAssemblies="@lazyLoadedAssemblies" OnNavigate="@OnNavigate">
+    <Found Context="routeData">
+        <RouteView RouteData="@routeData" />
+    </Found>
+    <NotFound>
+        <LayoutView Layout="@typeof(RouterTestLayout)">
+            <div id="test-info">Oops, that component wasn't found!</div>
+        </LayoutView>
+    </NotFound>
+</Router>
+
+@code {
+    private List<Assembly> lazyLoadedAssemblies = new List<Assembly>();
+
+    public bool OnNavigate (string uri)
+    {
+        Console.WriteLine($"Running OnNavigate for {uri}...");
+        _ = LoadAssemblies(uri);
+        return false;
+    }
+
+    private async Task LoadAssemblies(string uri)
+    {
+        bool loadedAssemblies = false;
+        try
+        {
+            if (uri.EndsWith("WithDynamicAssembly"))
+            {
+                loadedAssemblies = true;
+                Console.WriteLine($"Loading assemblies for WithDynamicAssembly...");
+                var assemblies = await lazyLoader.LoadAssembliesAsync(new List<string>(){ "Newtonsoft.Json.dll" });
+                lazyLoadedAssemblies.AddRange(assemblies);
+            }
+        }
+        catch (Exception e)
+        {
+            Console.WriteLine($"Error when loading assemblies: {e}");
+        }
+        finally {
+            if (loadedAssemblies)
+            {
+                Console.WriteLine("Calling StateHasChanged...");
+                StateHasChanged();
```

Comment:
It seems like the approach we're using to know when assemblies are loaded and ready to use is "whenever the `Router` next gets rendered". Am I misreading this, or is that the case?

My concern is that `StateHasChanged` might be executed before the assemblies are loaded for unrelated reasons. For example, maybe there's some other UI within the component holding the `<Router>`, and the user may interact with it.

If I'm completely misunderstanding the flow here then please let me know! Otherwise, perhaps rather than having `OnNavigate` return a `bool`, it should be returning a `Task`.

I'm also unsure what it would mean if a developer chooses to return `false` from `OnNavigate` and then never re-renders. Does that mean the URL is the address bar just no longer reflects what's in the UI? That seems like an odd situation and also makes me wonder if it would be better to return a `Task` so you can do async things, but don't get the option to say "don't navigate" (since you can't stop the URL from changing - it already has done).

Reply:
Also will you be adding something like a `Loading` render fragment parameter to the `Router` so the user can display some UI while waiting for the assemblies to be fetched?

Reply:
> I'm also unsure what it would mean if a developer chooses to return false from OnNavigate and then never re-renders. Does that mean the URL is the address bar just no longer reflects what's in the UI?

Correct. If the developer returns `false` but never calls `StateHasChanged`, they're in for a bad time because the component associated with the route will not be rendered.

> If I'm completely misunderstanding the flow here then please let me know! Otherwise, perhaps rather than having OnNavigate return a bool, it should be returning a Task.

Having `OnNavigate` return a Task would require changing the Router's internal implements (`NotifyLocationChanged`, etc) to support awaiting the task. It'll likely also require moving the logic for adding the lazy-loaded assemblies to the `AdditionalAssemblies` into the Router. While crufty, this approach strikes a balance with reducing the number of changes that need to be made to the router to support lazy-loading and gives users the flexibility to handle some aspects of this feature on their own (e.g. where to show the loading UI, when to re-render. etc).

Now, there very well might be scenarios where this API implementation doesn't work. With that in mind, I'd propose shipping preview7 support with this implementation, getting feedback on where the rough edges are, then refining as needed.

Let me know your thoughts on this.

Reply:
TBH if we think this API shape is crufty, I’d rather keep working towards one that we like. Including this in preview 7 doesn’t help us much if we’re going to change both the API and the intervals afterwards. Please don’t feel rushed on this - it’s completely reasonable to get this into preview 8 with the desired feature set.

Reply:
To clarify, I don't think the overall API shape is crufty. I expect that any changes we make will likely be additive to the `OnNavigate` handler. There are parts of this PR that are foundational, like the lazy loading service and it's JS components, that will probably not change moving forward and would be good to get in sooner than later.

> Also will you be adding something like a Loading render fragment parameter to the Router so the user can display some UI while waiting for the assemblies to be fetched?

With this model, users can implement the loading UI by setting a `loading` variable in their OnNavigate handler and conditionally rendering the loading UI based on the value of that property.

Reply:
Thanks for clarifying. If we have an API for which we think future changes are additive that sounds good - I just wanted to avoid dead ends. 

Do you think it would be possible to extend the test case to include displaying a “loading” state and representing what we think developers should do to use this fully? I have a slight suspicion that returning the bool might be unnecessary and void might be sufficient depending on what you render when in the loading state, but it would be good to find out for sure via a completed example.

Reply:
I added a test case for this. I'll also include this in the docs.

> I have a slight suspicion that returning the bool might be unnecessary and void might be sufficient depending on what you render when in the loading state, but it would be good to find out for sure via a completed example.

The suspicion is correct.